File: 240120/연속 부분 합의 최댓값 구하기/max-of-partial-sum.cpp
#include <iostream>
#include <vector>

using namespace std;

int main(){
    int n;
    cin>>n;

    vector<int> arr(100000);
    vector<int> dp(100000, -1001);


    for(int i=0; i<n; i++){
        cin >> arr[i];
    }

    dp[0] = arr[0];

    for(int i=1; i<n; i++){
        if(dp[i-1]+arr[i] > arr[i]) dp[i] = dp[i-1]+arr[i];
        else dp[i] = arr[i];
    }

    int max = -1001;
    for(int i=0; i<n; i++){
        max = max < dp[i] ? dp[i] : max;
    }
    cout << max;
}

File: 251109/연속 부분 합의 최댓값 구하기/max-of-partial-sum.py
N = int(input())
arr = list(map(int, input().split()))

# Please write your code here.
INT_MIN = (1<<20) * -1

dp = [INT_MIN for _ in range(len(arr))] # i가 마지막일 때 최대 
dp[0] = arr[0]

for i in range(1, len(dp)):
    dp[i] = max(dp[i-1] + arr[i], arr[i])

print(max(dp))
